### ISSUE TITLE
Feature request: ability to add arbitrary route with template / controller / variables from path (and document)

### ISSUE THREAD

**Issue by starsinmypockets** [datopian/frontend-v2#11]
## User Story
As a themer / developer, I want to add a custom page to the front end. 
I want to use a clear interface with reasonable defaults.

## Details
Maybe the interface looks something like:
* Add route config to routes.json:
```json
"routes": [
  {
    "title": "My Cool Page",
    "route": "/cool-page/%arg",
    "controller": "cool-page-controller.js",
    "template": "cool-page-template.html"  
  }
]
```
* add template file at `themes/MY-THEME/views/cool-page-template.html`
* add controller at `themes/MY-THEME/controllers/cool-page-controller.html`

## Notes
Not sure if something like this is in place already?  An alternative to static json / yaml type of config might be to search themes directories for routes.js file which could return additional express endpoint and which the application could then use to extend the existing express routes. This would allow developers to loop in custom pages, code, etc, without disrupting the base installation.

**Comment by starsinmypockets** [comment]
Doesn't look like I can assign myself but I'm working on a PR for this

**Comment by anuveyatsu** [comment]
@starsinmypockets what is your use case - can you please provide an example? E.g., if you need a custom static page which can be done in CMS, then you don't need to define custom route.

**Comment by smth** [comment]
This might be a separate issue, but could be resolved by the above, so will add as a comment here to begin with. Currently every "content" page uses the same template `static.html`. That means that if you want blog related stuff in or alongside your blog posts, then that stuff will also be on your "About" page.

I say it might be a separate issue, because there is an `about.html` template, which I would assume is to address this specific issue, but it's not being used, as far as I can tell.

**Comment by anuveyatsu** [comment]
@smth good catch - `about.html` wasn't used by `about` page. Currently, all static pages are using `static.html`.

To solve this issue I think we need to define which static pages need its own template. I guess for now it is only `about` page. In the future, we may need `contact` page as well. @starsinmypockets WDYT?

Options:

1. Create a separate router for `about` and other special static pages so that they use a custom template, e.g., `about.html`.
2. You can create some layout in WordPress by using the shortcode feature, e.g., I  can create a contact page - https://edscms.home.blog/contact/. However, we will need to do some tweaks to make it work.

**Comment by smth** [comment]
In the case of Open Data DK, "About" will be a collection of pages, rather than a single page. I would suggest that all pages in such a section have navigation in the template, linking to each page in that section.

**Comment by anuveyatsu** [comment]
@smth do you mean something like `/news` page where we have a list of posts/news returned from CMS backend or something more custom where you can hard code selected pages?

**Comment by starsinmypockets** [comment]
See https://github.com/datopian/frontend-v2/pull/12

**Comment by anuveyatsu** [comment]
@starsinmypockets could you give us an example of usage in your case? It would be useful to do the analysis.

**Comment by starsinmypockets** [comment]
We should provide hooks so that developers can add controllers, routes, and middleware, that way we are developing an unopinionated framework. The use case I am thinking of is the use case we didn't think of, whatever that may be :)

**Comment by starsinmypockets** [comment]
In my case (see the initial example) I wanted to add a route with a custom template that accepted an argument from the url.

**Comment by starsinmypockets** [comment]
The same way that ckan allows custom extensions, drupal, wordpress, etc have plugin formats, we should simply include a place where developers can plug in custom code / routes  / controllers, as a basic design feature of a flexible platform :) I don't think the "what will you do with it" question is really relevant. It's sort of like asking "if I add a USB port, what will you plug into it" -- the answer is, a peripheral! Anything I want:
- a temperature gauge
- a printer
- a hand-warmer
- whatever

The answer here is the same - if I can add a route or an extension I can:
- add a custom route, define route arguments, and define a custom template
- talk to a third-party api, preprocess data, and load a custom template with an embedded javascript app
- add an argument for language selection (this might already exist and should be included in the product, but this would allow me to accomplish this without)
- extend or override existing views (custom data view pages, for instance)
- apply logging or api analytics via express middleware
- anything I can do with a controller
- anything I can do with a template
- MVC development

**Comment by starsinmypockets** [comment]
A bit an aside on philosophy: As a react and an express developer, I understand how the technologies work and I don't want to use a _subset_ of the technologies (restrictive product model) I want to use a _superset_ (permissive library / framework / extension model). I want access to a set of stable and tested features (paths, pages, integrations, tests, etc) that already do things I frequently do (home page, dataset page, frictionless data integration, etc), and in addition I want to be able to do unrestricted test-driven development using the tools (feature development, third-party integrations, manipulate data, define endpoints, etc).

**Comment by starsinmypockets** [comment]
cc @rufuspollock

**Comment by rufuspollock** [comment]
@starsinmypockets i recommend analysing this properly in a hackmd with @anuveyatsu tomorrow (ie. set out the specific user story, do analysis of options for doing it etc)
